File: src/main/scala/Permutations.scala
import scala.collection.mutable.ListBuffer

object Permutations extends App {
  def perm(arr:Array[Int]): List[List[Int]] = {
    val result = new ListBuffer[List[Int]]()
    generate(arr, Array.ofDim[Boolean](arr.length), result, List.empty)
    result.toList
  }

  private def generate(s: Array[Int],  chosen:Array[Boolean], res: ListBuffer[List[Int]], buf: List[Int]):Unit = {
    if(buf.size == s.length) res += buf
    else {
      for(i <- 0 to s.length - 1) {
        if(!chosen(i)){
          val perm = buf :+ s(i)
          chosen(i) = true
          generate(s, chosen, res, perm)
          chosen(i) = false
        }
      }
    }
  }

   val arr = Array(1,2,3)
   val res = perm(arr)
   res.foreach(l => println(l.mkString(",")))

}

File: src/main/scala/Permutations.py
class Solution:
    def permutations(self, arr):
        if not arr:
            return

        def perm(l, r, res):
            if l == r:
                res.append(arr[::])
                return
            for i in range(l, r + 1):
                arr[i],arr[l] = arr[l],arr[i]
                perm(l+1, r, res)
                arr[i], arr[l] = arr[l], arr[i]
        res = []
        perm(0, len(arr) - 1, res)
        return res


sol = Solution()
print(sol.permutations([1, 2, 3, 4]))
